{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import statsmodels.stats.multitest as smt\n",
    "\n",
    "def sin_cos_24hr_encoder(df, col):\n",
    "    \"\"\"\n",
    "    Given a dataframe and column \n",
    "    This function takes a dataframe and a column that contains hours of the day,\n",
    "    and creates two new columns 'Hour_Sin' and 'Hour_Cos'. These columns contain\n",
    "    the sine and cosine encodings of the time, which is a common technique used\n",
    "    for cyclical features like time of day in machine learning models.\n",
    "\n",
    "    The sine and cosine values are calculated such that they form a circle when plotted,\n",
    "    which is useful for the model to recognize the cyclical nature of the data\n",
    "    (e.g., 23 hours is close to 0 hours).\n",
    "\n",
    "    Parameters:\n",
    "    df (DataFrame): The pandas dataframe to process.\n",
    "    col (str): The column name with the hour values (0 to 23).\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: The dataframe with two additional columns 'Hour_Sin' and 'Hour_Cos'.\n",
    "    \"\"\"\n",
    "    time_delta_hours_sin = np.sin(2 * np.pi * df[col] / 24)\n",
    "    time_delta_hours_cos = np.cos(2 * np.pi * df[col] / 24)\n",
    "    df['Hour_Sin'] = time_delta_hours_sin\n",
    "    df['Hour_Cos'] = time_delta_hours_cos\n",
    "\n",
    "    return df\n",
    "\n",
    "def sin_cos_24hr_decoder(df, sin_t, cos_t):\n",
    "    \"\"\"\n",
    "    This function takes a dataframe and the column names for sine and cosine\n",
    "    values that represent time. It calculates the corresponding hour of the day\n",
    "    for each row and creates a new column 'hour_int' with these values.\n",
    "    \n",
    "    Parameters:\n",
    "    df (DataFrame): The pandas dataframe to process.\n",
    "    sin_t (str): The column name for the sine of time values.\n",
    "    cos_t (str): The column name for the cosine of time values.\n",
    "    \n",
    "    Returns:\n",
    "    DataFrame: The dataframe with an additional 'hour_int' column.\n",
    "    \"\"\"\n",
    "    # Compute the angle in radians from the sin and cos values\n",
    "    angle_rad = np.arctan2(df[sin_t], df[cos_t])\n",
    "    \n",
    "   # Ensure the angle is positive for all values\n",
    "    angle_rad[angle_rad < 0] += 2 * np.pi\n",
    "    \n",
    "    # Convert the angle in radians to hours and add as a new column\n",
    "    df['hour_int'] = round(angle_rad / (2 * np.pi) * 24, 3)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def construct_cellbox_input(Perturbations, Molecular, Phenotype, Output_path):\n",
    "    \"\"\"\n",
    "    Load Perturbation, Molecular, and Phenotype data \n",
    "    Where the samples correspond to column names and rows correspond to nodes in the Cellbox output\n",
    "    \n",
    "    Parameters:\n",
    "        Perturbation: Changing experimental conditions \n",
    "        Molecular: Measured cellular response \n",
    "        Phenotype: Measured output corresponding to phenotype of interest\n",
    "        Output_Directory: Where the results should be written\n",
    "        \n",
    "    Returns:\n",
    "        expression_matrix: a matrix of values corresponding to the indivual input values compiled for cellboxs\n",
    "        perturbation_matrix: a matrix of values corresponding to initial base condition values (0) and associated perturbations\n",
    "        node_index: the names of all individual nodes (rownames)\n",
    "        \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    experiment = Output_path.split(\"/\")[-2]\n",
    "    n_protein_nodes = Molecular.shape[0]\n",
    "    n_activity_nodes = n_protein_nodes + Phenotype.shape[0]\n",
    "    n_x = n_activity_nodes + Perturbations.shape[0]\n",
    "\n",
    "    config_json = construct_cellbox_json(experiment, n_protein_nodes, n_activity_nodes, n_x)\n",
    "\n",
    "    activity_nodes = pd.merge(Molecular.T, Phenotype.T, left_index=True, right_index=True)\n",
    "    expr_matrix = pd.merge(activity_nodes, Perturbations.T, left_index=True, right_index=True)\n",
    "\n",
    "    \n",
    "    #expr_matrix = pd.concat([Molecular, Phenotype, Perturbations]).T\n",
    "    # Construct a dataframe where all initial values are zero in the shape of (# perturbations + # phenotypes, # samples)\n",
    "    # Add corrresponding perturbation data\n",
    "    pert_matrix = np.concatenate([np.zeros((Molecular.shape[0]+Phenotype.shape[0], Molecular.shape[1])),\n",
    "                                  Perturbations]).T\n",
    "    \n",
    "    node_index = expr_matrix.columns.to_series()\n",
    "    sample_order = expr_matrix.index.to_series()\n",
    "    \n",
    "    os.makedirs(Output_path, exist_ok=True)\n",
    "\n",
    "    with open(os.path.join(Output_path, 'config.json'), 'w') as json_file:\n",
    "        json.dump(config_json, json_file)\n",
    "\n",
    "    np.savetxt(os.path.join(Output_path, 'expr_matr.csv'), expr_matrix.to_numpy(), delimiter=\",\")\n",
    "    np.savetxt(os.path.join(Output_path, 'pert_matr.csv'), pert_matrix, delimiter=\",\")\n",
    "    node_index.to_csv(os.path.join(Output_path, 'node_index.csv'), header=False, index=False)\n",
    "    sample_order.to_csv(os.path.join(Output_path, 'sample_order.csv'), header=False, index=False)\n",
    "\n",
    "def construct_cellbox_json(experiment, n_protein_nodes, n_activity_nodes, n_x):\n",
    "    \"\"\"\n",
    "    Constructs a JSON object for the Cellbox experiment with the given parameters.\n",
    "\n",
    "    Parameters:\n",
    "    path (str): The base path for the files.\n",
    "    n_protein_nodes (int): The number of protein nodes.\n",
    "    n_activity_nodes (int): The number of activity nodes.\n",
    "    n_x (int): The number of x nodes.\n",
    "\n",
    "    Returns:\n",
    "    dict: The constructed JSON object.\n",
    "    \"\"\"\n",
    "    return {\n",
    "        \"experiment_id\": experiment,\n",
    "        \"experiment_type\": \"random partition\",\n",
    "        \"model\": \"CellBox\",\n",
    "        \"sparse_data\": False,\n",
    "        \"pert_file\": f\"Execution_files/{experiment}/pert_matr.csv\",\n",
    "        \"expr_file\": f\"Execution_files/{experiment}/expr_matr.csv\",\n",
    "        \"node_index_file\": f\"Execution_files/{experiment}/node_index.csv\",\n",
    "        \"n_protein_nodes\": n_protein_nodes,\n",
    "        \"n_activity_nodes\": n_activity_nodes,\n",
    "        \"n_x\": n_x,\n",
    "        \"trainset_ratio\": 0.7,\n",
    "        \"validset_ratio\": 0.8,\n",
    "        \"batchsize\": 32,\n",
    "        \"add_noise_level\": 0,\n",
    "        \"envelop_form\": \"tanh\",\n",
    "        \"dT\": 0.1,\n",
    "        \"envelop\": 0,\n",
    "        \"ode_degree\": 1,\n",
    "        \"ode_solver\": \"heun\",\n",
    "        \"ode_last_steps\": 2,\n",
    "        \"l1lambda\": 1e-4,\n",
    "        \"l2lambda\": 1e-4,\n",
    "        \"n_epoch\": 10000,\n",
    "        \"n_iter\": 10000,\n",
    "        \"n_iter_buffer\": 50,\n",
    "        \"n_iter_patience\": 100,\n",
    "        \"stages\": [\n",
    "            {\n",
    "                \"nT\": 100,\n",
    "                \"sub_stages\": [\n",
    "                    {\"lr_val\": 0.1, \"l1lambda\": 0.01, \"n_iter_patience\": 1000},\n",
    "                    {\"lr_val\": 0.01, \"l1lambda\": 0.01},\n",
    "                    {\"lr_val\": 0.01, \"l1lambda\": 0.0001},\n",
    "                    {\"lr_val\": 0.001, \"l1lambda\": 0.00001}\n",
    "                ]\n",
    "            },\n",
    "            {\n",
    "                \"nT\": 200,\n",
    "                \"sub_stages\": [\n",
    "                    {\"lr_val\": 0.001, \"l1lambda\": 0.0001}\n",
    "                ]\n",
    "            },\n",
    "            {\n",
    "                \"nT\": 400,\n",
    "                \"sub_stages\": [\n",
    "                    {\"lr_val\": 0.001, \"l1lambda\": 0.0001}\n",
    "                ]\n",
    "            }\n",
    "        ],\n",
    "        \"export_verbose\": 3,\n",
    "        \"ckpt_name\": experiment\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "A = pd.read_csv(\"../Data/ica/A.csv\", index_col=0)\n",
    "log2FC  = pd.read_csv(\"../Data/deseq2/Log2FC_All_Expression.csv\", index_col=0)\n",
    "logTPM_norm = pd.read_csv(\"../Data/ica/logTPM_counts_normalized_passingQC.csv\", index_col=0)\n",
    "tf_order=pd.read_csv('../Data/deseq2/tf_order.csv',header=None)\n",
    "# From Puszynska & O'Shea 2017 (DOI: https://doi.org/10.7554/eLife.23210)\n",
    "light_pert = pd.read_csv(\"../Data/literature_metadata/circadian_time_light_Puszynska_2017.csv\")\n",
    "light_pert = sin_cos_24hr_encoder(light_pert, 'Circadian_time_h')\n",
    "\n",
    "## Load the genome annotation\n",
    "ann = pd.read_csv(\"../Data/annotation/Selongatus_PCC7942_ref_tfs.csv\")\n",
    "ann.set_index('locus_tag', inplace=True)\n",
    "\n",
    "# Load all predicted TFs\n",
    "pred_TFs = pd.read_csv(\"../Data/annotation/All_predicted_TFs.csv\")\n",
    "\n",
    "# Load all known regulatory interactions\n",
    "regs = pd.read_csv(\"../Data/annotation/TRN_complete.csv\", index_col=0)\n",
    "\n",
    "metadata = pd.read_csv('../Data/metadata/Metadata_Perturbation_Passing_QC.csv', index_col=0)\n",
    "metadata.rename(columns={\"Project_tag\":\"project\", \"Condition\":\"condition\"}, inplace=True)\n",
    "metadata['Ref_condition'] = metadata['project'] + \"_\" + metadata['Ref_condition']\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "markson_ann = pd.read_excel(\"../Data/literature_metadata/CircadianGenes_Markson_2013.xlsm\", skiprows=1)\n",
    "circadian_genes = markson_ann[markson_ann['Reproducibly circadian'] == 1]\n",
    "print(f'Number of circadian genes: {circadian_genes.shape[0]}')\n",
    "#circadian_genes = circadian_genes[~(circadian_genes['Chromosome or plasmid'] == 'pANS')]\n",
    "circadian_genes.head(10)\n",
    "\n",
    "## Manual mapping of circadian genes on PANL to alternative annotation using uniprot & Biocyc\n",
    "## Single pANS gene not included\n",
    "map_dict = {'HKK26_RS00110': 'Synpcc7942_B2626',\n",
    "            'HKK26_RS00115': 'Synpcc7942_B2637',\n",
    "            'HKK26_RS00120': 'Synpcc7942_B2648',\n",
    "            'HKK26_RS00190': 'Synpcc7942_B2623',\n",
    "            'HKK26_RS00235': 'Synpcc7942_B2655', \n",
    "            'HKK26_RS00240': 'Synpcc7942_B2632',\n",
    "            'HKK26_RS00020': 'Synpcc7942_B2635'}\n",
    "\n",
    "circadian_genes_up = ann[(ann.old_locus_tag.isin(circadian_genes['Gene ID'])) | (ann.index.isin(map_dict.keys()))]\n",
    "print(f\"Number of genes mapped to annotation: {circadian_genes_up.shape[0]}\")\n",
    "circadian_genes_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load difference in gene expression data for TFs (1. Input into ICA)\n",
    "light_pert_samples = light_pert[['SRA', 'Light', 'Hour_Sin', 'Hour_Cos']].set_index('SRA').T\n",
    "light_pert_samples\n",
    "\n",
    "oshea_samples_logTPM_norm_TFs = logTPM_norm.loc[pred_TFs.TFs, light_pert_samples.columns]\n",
    "oshea_samples_logTPM_norm_TFs.index = oshea_samples_logTPM_norm_TFs.index.map('TF_{}'.format)\n",
    "oshea_samples_logTPM_norm_TFs\n",
    "\n",
    "light_pert_conditions = light_pert[['Condition_pi', 'Light', 'Hour_Sin', 'Hour_Cos']\n",
    "                                  ].drop_duplicates().set_index('Condition_pi').T.drop(columns='wt_clearday_dawn_0.5h')\n",
    "\n",
    "#Load log2FC data (2. Output from DESeq2)\n",
    "oshea_samples_log2FC_TFs = log2FC.loc[pred_TFs.TFs, light_pert_conditions.columns]\n",
    "oshea_samples_log2FC_TFs.index = oshea_samples_log2FC_TFs.index.map('TF_{}'.format)\n",
    "\n",
    "# Modulon response by sample\n",
    "oshea_IC_sample = A.loc[:, light_pert_samples.columns]\n",
    "\n",
    "## Average the modulon resposne across conditions \n",
    "## wt_clearday_dawn_0.5h removed as it was used as the base condition for all differential expression comparisons\n",
    "\n",
    "avg_A = []\n",
    "for sample, group in light_pert.groupby('Condition_pi'):\n",
    "    avg = oshea_IC_sample[group.SRA].mean(axis=1)\n",
    "    avg_A.append(avg)\n",
    "    \n",
    "avg_A_df = pd.concat(avg_A, axis=1)\n",
    "avg_A_df.columns = light_pert.groupby('Condition_pi').groups.keys()\n",
    "avg_A_df = avg_A_df.drop('wt_clearday_dawn_0.5h', axis=1)\n",
    "avg_A_df = avg_A_df.reindex(light_pert_conditions.columns.values, axis=1)\n",
    "avg_A_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "# Module model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "circadian_tfs = set(circadian_genes_up.index).intersection(set(pred_TFs.TFs))\n",
    "add_regs = ['RpaA', 'RpaB', 'rre1', 'HrcA', 'NtcA']\n",
    "add_regs_locus_tags = regs[regs.Regulator_Name.isin(add_regs)].regulatoryGene.unique()\n",
    "\n",
    "\n",
    "print(f'Known Regulators identified in circadian phase: {regs[regs.regulatoryGene.isin(circadian_tfs)].Regulator_Name.unique()}')\n",
    "print(f'Inclusion of known regulators: {regs[regs.regulatoryGene.isin(circadian_tfs.union(set(add_regs_locus_tags)))].Regulator_Name.unique()}')\n",
    "\n",
    "circ_log2FC_TFs = oshea_samples_log2FC_TFs.loc[list([\"TF_\" + TF for TF in circadian_tfs.union(set(add_regs_locus_tags))]),:]\n",
    "circ_log2FC_TFs = circ_log2FC_TFs.reindex(tf_order[0].values)\n",
    "construct_cellbox_input(light_pert_conditions, circ_log2FC_TFs, avg_A_df, '../Execution_files/Cellbox_Module')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "# Gene model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "circ_log2FC_genes = log2FC[log2FC.index.isin(circadian_genes_up.index)]\n",
    "\n",
    "construct_cellbox_input(light_pert_conditions, circ_log2FC_TFs, circ_log2FC_genes, \n",
    "                        \"../Execution_files/Cellbox_Gene\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pimlomics_conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
