{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d808a10",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import statsmodels.stats.multitest as smt\n",
    "\n",
    "def sin_cos_24hr_encoder(df, col):\n",
    "    \"\"\"\n",
    "    Given a dataframe and column \n",
    "    This function takes a dataframe and a column that contains hours of the day,\n",
    "    and creates two new columns 'Hour_Sin' and 'Hour_Cos'. These columns contain\n",
    "    the sine and cosine encodings of the time, which is a common technique used\n",
    "    for cyclical features like time of day in machine learning models.\n",
    "\n",
    "    The sine and cosine values are calculated such that they form a circle when plotted,\n",
    "    which is useful for the model to recognize the cyclical nature of the data\n",
    "    (e.g., 23 hours is close to 0 hours).\n",
    "\n",
    "    Parameters:\n",
    "    df (DataFrame): The pandas dataframe to process.\n",
    "    col (str): The column name with the hour values (0 to 23).\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: The dataframe with two additional columns 'Hour_Sin' and 'Hour_Cos'.\n",
    "    \"\"\"\n",
    "    time_delta_hours_sin = np.sin(2 * np.pi * df[col] / 24)\n",
    "    time_delta_hours_cos = np.cos(2 * np.pi * df[col] / 24)\n",
    "    df['Hour_Sin'] = time_delta_hours_sin\n",
    "    df['Hour_Cos'] = time_delta_hours_cos\n",
    "\n",
    "    return df\n",
    "\n",
    "def sin_cos_24hr_decoder(df, sin_t, cos_t):\n",
    "    \"\"\"\n",
    "    This function takes a dataframe and the column names for sine and cosine\n",
    "    values that represent time. It calculates the corresponding hour of the day\n",
    "    for each row and creates a new column 'hour_int' with these values.\n",
    "    \n",
    "    Parameters:\n",
    "    df (DataFrame): The pandas dataframe to process.\n",
    "    sin_t (str): The column name for the sine of time values.\n",
    "    cos_t (str): The column name for the cosine of time values.\n",
    "    \n",
    "    Returns:\n",
    "    DataFrame: The dataframe with an additional 'hour_int' column.\n",
    "    \"\"\"\n",
    "    # Compute the angle in radians from the sin and cos values\n",
    "    angle_rad = np.arctan2(df[sin_t], df[cos_t])\n",
    "    \n",
    "   # Ensure the angle is positive for all values\n",
    "    angle_rad[angle_rad < 0] += 2 * np.pi\n",
    "    \n",
    "    # Convert the angle in radians to hours and add as a new column\n",
    "    df['hour_int'] = round(angle_rad / (2 * np.pi) * 24, 3)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def construct_cellbox_input(Perturbations, Molecular, Phenotype, Output_path):\n",
    "    \"\"\"\n",
    "    Load Perturbation, Molecular, and Phenotype data \n",
    "    Where the samples correspond to column names and rows correspond to nodes in the Cellbox output\n",
    "    \n",
    "    Parameters:\n",
    "        Perturbation: Changing experimental conditions \n",
    "        Molecular: Measured cellular response \n",
    "        Phenotype: Measured output corresponding to phenotype of interest\n",
    "        Output_Directory: Where the results should be written\n",
    "        \n",
    "    Returns:\n",
    "        expression_matrix: a matrix of values corresponding to the indivual input values compiled for cellboxs\n",
    "        perturbation_matrix: a matrix of values corresponding to initial base condition values (0) and associated perturbations\n",
    "        node_index: the names of all individual nodes (rownames)\n",
    "        \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    experiment = Output_path.split(\"/\")[-2]\n",
    "    n_protein_nodes = Molecular.shape[0]\n",
    "    n_activity_nodes = n_protein_nodes + Phenotype.shape[0]\n",
    "    n_x = n_activity_nodes + Perturbations.shape[0]\n",
    "\n",
    "    config_json = construct_cellbox_json(experiment, n_protein_nodes, n_activity_nodes, n_x)\n",
    "\n",
    "    activity_nodes = pd.merge(Molecular.T, Phenotype.T, left_index=True, right_index=True)\n",
    "    expr_matrix = pd.merge(activity_nodes, Perturbations.T, left_index=True, right_index=True)\n",
    "\n",
    "    \n",
    "    #expr_matrix = pd.concat([Molecular, Phenotype, Perturbations]).T\n",
    "    # Construct a dataframe where all initial values are zero in the shape of (# perturbations + # phenotypes, # samples)\n",
    "    # Add corrresponding perturbation data\n",
    "    pert_matrix = np.concatenate([np.zeros((Molecular.shape[0]+Phenotype.shape[0], Molecular.shape[1])),\n",
    "                                  Perturbations]).T\n",
    "    \n",
    "    node_index = expr_matrix.columns.to_series()\n",
    "    sample_order = expr_matrix.index.to_series()\n",
    "    \n",
    "    os.makedirs(Output_path, exist_ok=True)\n",
    "\n",
    "    with open(os.path.join(Output_path, 'config.json'), 'w') as json_file:\n",
    "        json.dump(config_json, json_file)\n",
    "\n",
    "    np.savetxt(os.path.join(Output_path, 'expr_matr.csv'), expr_matrix.to_numpy(), delimiter=\",\")\n",
    "    np.savetxt(os.path.join(Output_path, 'pert_matr.csv'), pert_matrix, delimiter=\",\")\n",
    "    node_index.to_csv(os.path.join(Output_path, 'node_index.csv'), header=False, index=False)\n",
    "    sample_order.to_csv(os.path.join(Output_path, 'sample_order.csv'), header=False, index=False)\n",
    "\n",
    "def construct_cellbox_json(experiment, n_protein_nodes, n_activity_nodes, n_x):\n",
    "    \"\"\"\n",
    "    Constructs a JSON object for the Cellbox experiment with the given parameters.\n",
    "\n",
    "    Parameters:\n",
    "    path (str): The base path for the files.\n",
    "    n_protein_nodes (int): The number of protein nodes.\n",
    "    n_activity_nodes (int): The number of activity nodes.\n",
    "    n_x (int): The number of x nodes.\n",
    "\n",
    "    Returns:\n",
    "    dict: The constructed JSON object.\n",
    "    \"\"\"\n",
    "    return {\n",
    "        \"experiment_id\": experiment,\n",
    "        \"experiment_type\": \"random partition\",\n",
    "        \"model\": \"CellBox\",\n",
    "        \"sparse_data\": False,\n",
    "        \"pert_file\": f\"Execution_files/{experiment}/pert_matr.csv\",\n",
    "        \"expr_file\": f\"Execution_files/{experiment}/expr_matr.csv\",\n",
    "        \"node_index_file\": f\"Execution_files/{experiment}/node_index.csv\",\n",
    "        \"n_protein_nodes\": n_protein_nodes,\n",
    "        \"n_activity_nodes\": n_activity_nodes,\n",
    "        \"n_x\": n_x,\n",
    "        \"trainset_ratio\": 0.7,\n",
    "        \"validset_ratio\": 0.8,\n",
    "        \"batchsize\": 32,\n",
    "        \"add_noise_level\": 0,\n",
    "        \"envelop_form\": \"tanh\",\n",
    "        \"dT\": 0.1,\n",
    "        \"envelop\": 0,\n",
    "        \"ode_degree\": 1,\n",
    "        \"ode_solver\": \"heun\",\n",
    "        \"ode_last_steps\": 2,\n",
    "        \"l1lambda\": 1e-4,\n",
    "        \"l2lambda\": 1e-4,\n",
    "        \"n_epoch\": 10000,\n",
    "        \"n_iter\": 10000,\n",
    "        \"n_iter_buffer\": 50,\n",
    "        \"n_iter_patience\": 100,\n",
    "        \"stages\": [\n",
    "            {\n",
    "                \"nT\": 100,\n",
    "                \"sub_stages\": [\n",
    "                    {\"lr_val\": 0.1, \"l1lambda\": 0.01, \"n_iter_patience\": 1000},\n",
    "                    {\"lr_val\": 0.01, \"l1lambda\": 0.01},\n",
    "                    {\"lr_val\": 0.01, \"l1lambda\": 0.0001},\n",
    "                    {\"lr_val\": 0.001, \"l1lambda\": 0.00001}\n",
    "                ]\n",
    "            },\n",
    "            {\n",
    "                \"nT\": 200,\n",
    "                \"sub_stages\": [\n",
    "                    {\"lr_val\": 0.001, \"l1lambda\": 0.0001}\n",
    "                ]\n",
    "            },\n",
    "            {\n",
    "                \"nT\": 400,\n",
    "                \"sub_stages\": [\n",
    "                    {\"lr_val\": 0.001, \"l1lambda\": 0.0001}\n",
    "                ]\n",
    "            }\n",
    "        ],\n",
    "        \"export_verbose\": 3,\n",
    "        \"ckpt_name\": experiment\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "9d673a94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Biosample</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Bases</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>create_date</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Experiment.1</th>\n",
       "      <th>condition</th>\n",
       "      <th>...</th>\n",
       "      <th>pH</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Time_delta</th>\n",
       "      <th>Condition_pi</th>\n",
       "      <th>Project_condition_pi</th>\n",
       "      <th>Passed_Min_Reads</th>\n",
       "      <th>Passed_Global_Corr</th>\n",
       "      <th>passed_replicate_corr</th>\n",
       "      <th>passed_timeseries_corr</th>\n",
       "      <th>Overall_pass</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR807396</th>\n",
       "      <td>PRJNA196229</td>\n",
       "      <td>SAMN01999000</td>\n",
       "      <td>SRX259777</td>\n",
       "      <td>4.05 G</td>\n",
       "      <td>2.64 Gb</td>\n",
       "      <td>2015-11-15</td>\n",
       "      <td>Synechococcus elongatus PCC 7942 = FACHB-805</td>\n",
       "      <td>WT</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 days 04:00:00</td>\n",
       "      <td>WT_100h</td>\n",
       "      <td>FFA_WT_100h</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR807397</th>\n",
       "      <td>PRJNA196229</td>\n",
       "      <td>SAMN01999001</td>\n",
       "      <td>SRX259778</td>\n",
       "      <td>2.71 G</td>\n",
       "      <td>1.77 Gb</td>\n",
       "      <td>2015-11-15</td>\n",
       "      <td>Synechococcus elongatus PCC 7942 = FACHB-805</td>\n",
       "      <td>WT</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 days 04:00:00</td>\n",
       "      <td>WT_100h</td>\n",
       "      <td>FFA_WT_100h</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR807398</th>\n",
       "      <td>PRJNA196229</td>\n",
       "      <td>SAMN01999002</td>\n",
       "      <td>SRX259779</td>\n",
       "      <td>2.90 G</td>\n",
       "      <td>1.91 Gb</td>\n",
       "      <td>2015-11-15</td>\n",
       "      <td>Synechococcus elongatus PCC 7942 = FACHB-805</td>\n",
       "      <td>WT</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 days 04:00:00</td>\n",
       "      <td>WT_100h</td>\n",
       "      <td>FFA_WT_100h</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR807399</th>\n",
       "      <td>PRJNA196229</td>\n",
       "      <td>SAMN01999003</td>\n",
       "      <td>SRX259780</td>\n",
       "      <td>3.19 G</td>\n",
       "      <td>2.08 Gb</td>\n",
       "      <td>2015-11-15</td>\n",
       "      <td>Synechococcus elongatus PCC 7942 = FACHB-805</td>\n",
       "      <td>WT</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10 days 00:00:00</td>\n",
       "      <td>WT_240h</td>\n",
       "      <td>FFA_WT_240h</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR807400</th>\n",
       "      <td>PRJNA196229</td>\n",
       "      <td>SAMN01999004</td>\n",
       "      <td>SRX259781</td>\n",
       "      <td>1.67 G</td>\n",
       "      <td>1.09 Gb</td>\n",
       "      <td>2015-11-15</td>\n",
       "      <td>Synechococcus elongatus PCC 7942 = FACHB-805</td>\n",
       "      <td>WT</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10 days 00:00:00</td>\n",
       "      <td>WT_240h</td>\n",
       "      <td>FFA_WT_240h</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            BioProject     Biosample Experiment   Bases    Bytes create_date  \\\n",
       "SRA                                                                            \n",
       "SRR807396  PRJNA196229  SAMN01999000  SRX259777  4.05 G  2.64 Gb  2015-11-15   \n",
       "SRR807397  PRJNA196229  SAMN01999001  SRX259778  2.71 G  1.77 Gb  2015-11-15   \n",
       "SRR807398  PRJNA196229  SAMN01999002  SRX259779  2.90 G  1.91 Gb  2015-11-15   \n",
       "SRR807399  PRJNA196229  SAMN01999003  SRX259780  3.19 G  2.08 Gb  2015-11-15   \n",
       "SRR807400  PRJNA196229  SAMN01999004  SRX259781  1.67 G  1.09 Gb  2015-11-15   \n",
       "\n",
       "                                                 Strain Genotype Experiment.1  \\\n",
       "SRA                                                                             \n",
       "SRR807396  Synechococcus elongatus PCC 7942 = FACHB-805       WT      RNA-seq   \n",
       "SRR807397  Synechococcus elongatus PCC 7942 = FACHB-805       WT      RNA-seq   \n",
       "SRR807398  Synechococcus elongatus PCC 7942 = FACHB-805       WT      RNA-seq   \n",
       "SRR807399  Synechococcus elongatus PCC 7942 = FACHB-805       WT      RNA-seq   \n",
       "SRR807400  Synechococcus elongatus PCC 7942 = FACHB-805       WT      RNA-seq   \n",
       "\n",
       "          condition  ...  pH  Notes        Time_delta  Condition_pi  \\\n",
       "SRA                  ...                                              \n",
       "SRR807396        WT  ... NaN    NaN   4 days 04:00:00       WT_100h   \n",
       "SRR807397        WT  ... NaN    NaN   4 days 04:00:00       WT_100h   \n",
       "SRR807398        WT  ... NaN    NaN   4 days 04:00:00       WT_100h   \n",
       "SRR807399        WT  ... NaN    NaN  10 days 00:00:00       WT_240h   \n",
       "SRR807400        WT  ... NaN    NaN  10 days 00:00:00       WT_240h   \n",
       "\n",
       "           Project_condition_pi  Passed_Min_Reads  Passed_Global_Corr  \\\n",
       "SRA                                                                     \n",
       "SRR807396           FFA_WT_100h              True                True   \n",
       "SRR807397           FFA_WT_100h              True                True   \n",
       "SRR807398           FFA_WT_100h              True                True   \n",
       "SRR807399           FFA_WT_240h              True                True   \n",
       "SRR807400           FFA_WT_240h              True                True   \n",
       "\n",
       "           passed_replicate_corr  passed_timeseries_corr  Overall_pass  \n",
       "SRA                                                                     \n",
       "SRR807396                   True                   False          True  \n",
       "SRR807397                   True                   False          True  \n",
       "SRR807398                   True                   False          True  \n",
       "SRR807399                   True                   False          True  \n",
       "SRR807400                   True                   False          True  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = pd.read_csv(\"../Data/ica/A.csv\", index_col=0)\n",
    "log2FC  = pd.read_csv(\"../Data/deseq2/Log2FC_All_Expression.csv\", index_col=0)\n",
    "logTPM_norm = pd.read_csv(\"../Data/ica/logTPM_counts_normalized_passingQC.csv\", index_col=0)\n",
    "tf_order=pd.read_csv('../Data/deseq2/tf_order.csv',header=None)\n",
    "# From Puszynska & O'Shea 2017 (DOI: https://doi.org/10.7554/eLife.23210)\n",
    "light_pert = pd.read_csv(\"../Data/literature_metadata/circadian_time_light_Puszynska_2017.csv\")\n",
    "light_pert = sin_cos_24hr_encoder(light_pert, 'Circadian_time_h')\n",
    "\n",
    "## Load the genome annotation\n",
    "ann = pd.read_csv(\"Selongatus_PCC7942_ref_tfs.csv\")\n",
    "ann.set_index('locus_tag', inplace=True)\n",
    "\n",
    "# Load all predicted TFs\n",
    "pred_TFs = pd.read_csv(\"../Data/annotation/All_predicted_TFs.csv\")\n",
    "\n",
    "# Load all known regulatory interactions\n",
    "regs = pd.read_csv(\"TRN_complete.csv\", index_col=0)\n",
    "\n",
    "metadata = pd.read_csv('Metadata_Perturbation_Passing_QC.csv', index_col=0)\n",
    "metadata.rename(columns={\"Project_tag\":\"project\", \"Condition\":\"condition\"}, inplace=True)\n",
    "metadata['Ref_condition'] = metadata['project'] + \"_\" + metadata['Ref_condition']\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d6e6003c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of circadian genes: 856\n",
      "Number of genes mapped to annotation: 846\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/john002/miniconda3/envs/pimlomics_conda/lib/python3.12/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>ncbi_protein</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>Eggnog_description</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>...</th>\n",
       "      <th>ENTRAF_TF</th>\n",
       "      <th>ENTRAF_hmm_acc</th>\n",
       "      <th>ENTRAF_hmm_name</th>\n",
       "      <th>ENTRAF_Ortholog</th>\n",
       "      <th>DeepTFactor_TF</th>\n",
       "      <th>DeepTFactor_score</th>\n",
       "      <th>P2TF_TF</th>\n",
       "      <th>P2TF_class</th>\n",
       "      <th>P2TF_type</th>\n",
       "      <th>P2TF_description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00010</th>\n",
       "      <td>1262</td>\n",
       "      <td>2134</td>\n",
       "      <td>873</td>\n",
       "      <td>+</td>\n",
       "      <td>Synpcc7942_0002</td>\n",
       "      <td>WP_011243805.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>PFAM PRC-barrel domain</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0282</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00020</th>\n",
       "      <td>4596</td>\n",
       "      <td>6077</td>\n",
       "      <td>1482</td>\n",
       "      <td>+</td>\n",
       "      <td>Synpcc7942_0004</td>\n",
       "      <td>WP_011243803.1</td>\n",
       "      <td>purF</td>\n",
       "      <td>amidophosphoribosyltransferase</td>\n",
       "      <td>Catalyzes the formation of phosphoribosylamine...</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00025</th>\n",
       "      <td>6111</td>\n",
       "      <td>7706</td>\n",
       "      <td>1596</td>\n",
       "      <td>-</td>\n",
       "      <td>Synpcc7942_0005</td>\n",
       "      <td>WP_011377397.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>permease</td>\n",
       "      <td>secondary active sulfate transmembrane transpo...</td>\n",
       "      <td>P</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0069</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00035</th>\n",
       "      <td>8189</td>\n",
       "      <td>8539</td>\n",
       "      <td>351</td>\n",
       "      <td>-</td>\n",
       "      <td>Synpcc7942_0007</td>\n",
       "      <td>WP_011377398.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YtxH domain-containing protein</td>\n",
       "      <td>YtxH-like protein</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0110</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00040</th>\n",
       "      <td>8536</td>\n",
       "      <td>8859</td>\n",
       "      <td>324</td>\n",
       "      <td>-</td>\n",
       "      <td>Synpcc7942_0008</td>\n",
       "      <td>WP_011377399.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    start   end  length strand    old_locus_tag  \\\n",
       "locus_tag                                                         \n",
       "SYNPCC7942_RS00010   1262  2134     873      +  Synpcc7942_0002   \n",
       "SYNPCC7942_RS00020   4596  6077    1482      +  Synpcc7942_0004   \n",
       "SYNPCC7942_RS00025   6111  7706    1596      -  Synpcc7942_0005   \n",
       "SYNPCC7942_RS00035   8189  8539     351      -  Synpcc7942_0007   \n",
       "SYNPCC7942_RS00040   8536  8859     324      -  Synpcc7942_0008   \n",
       "\n",
       "                      ncbi_protein gene_name                    gene_product  \\\n",
       "locus_tag                                                                      \n",
       "SYNPCC7942_RS00010  WP_011243805.1       NaN            hypothetical protein   \n",
       "SYNPCC7942_RS00020  WP_011243803.1      purF  amidophosphoribosyltransferase   \n",
       "SYNPCC7942_RS00025  WP_011377397.1       NaN                        permease   \n",
       "SYNPCC7942_RS00035  WP_011377398.1       NaN  YtxH domain-containing protein   \n",
       "SYNPCC7942_RS00040  WP_011377399.1       NaN            hypothetical protein   \n",
       "\n",
       "                                                   Eggnog_description  \\\n",
       "locus_tag                                                               \n",
       "SYNPCC7942_RS00010                             PFAM PRC-barrel domain   \n",
       "SYNPCC7942_RS00020  Catalyzes the formation of phosphoribosylamine...   \n",
       "SYNPCC7942_RS00025  secondary active sulfate transmembrane transpo...   \n",
       "SYNPCC7942_RS00035                                  YtxH-like protein   \n",
       "SYNPCC7942_RS00040                                                NaN   \n",
       "\n",
       "                   COG_category  ... ENTRAF_TF ENTRAF_hmm_acc ENTRAF_hmm_name  \\\n",
       "locus_tag                        ...                                            \n",
       "SYNPCC7942_RS00010            S  ...     False            NaN             NaN   \n",
       "SYNPCC7942_RS00020            F  ...     False            NaN             NaN   \n",
       "SYNPCC7942_RS00025            P  ...     False            NaN             NaN   \n",
       "SYNPCC7942_RS00035            S  ...     False            NaN             NaN   \n",
       "SYNPCC7942_RS00040            X  ...     False            NaN             NaN   \n",
       "\n",
       "                   ENTRAF_Ortholog DeepTFactor_TF DeepTFactor_score P2TF_TF  \\\n",
       "locus_tag                                                                     \n",
       "SYNPCC7942_RS00010             NaN          False            0.0282   False   \n",
       "SYNPCC7942_RS00020             NaN          False            0.0137   False   \n",
       "SYNPCC7942_RS00025             NaN          False            0.0069   False   \n",
       "SYNPCC7942_RS00035             NaN          False            0.0110   False   \n",
       "SYNPCC7942_RS00040             NaN          False            0.0096   False   \n",
       "\n",
       "                   P2TF_class P2TF_type P2TF_description  \n",
       "locus_tag                                                 \n",
       "SYNPCC7942_RS00010        NaN       NaN              NaN  \n",
       "SYNPCC7942_RS00020        NaN       NaN              NaN  \n",
       "SYNPCC7942_RS00025        NaN       NaN              NaN  \n",
       "SYNPCC7942_RS00035        NaN       NaN              NaN  \n",
       "SYNPCC7942_RS00040        NaN       NaN              NaN  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markson_ann = pd.read_excel(\"../Data/literature_metadata/CircadianGenes_Markson_2013.xlsm\", skiprows=1)\n",
    "circadian_genes = markson_ann[markson_ann['Reproducibly circadian'] == 1]\n",
    "print(f'Number of circadian genes: {circadian_genes.shape[0]}')\n",
    "#circadian_genes = circadian_genes[~(circadian_genes['Chromosome or plasmid'] == 'pANS')]\n",
    "circadian_genes.head(10)\n",
    "\n",
    "## Manual mapping of circadian genes on PANL to alternative annotation using uniprot & Biocyc\n",
    "## Single pANS gene not included\n",
    "map_dict = {'HKK26_RS00110': 'Synpcc7942_B2626',\n",
    "            'HKK26_RS00115': 'Synpcc7942_B2637',\n",
    "            'HKK26_RS00120': 'Synpcc7942_B2648',\n",
    "            'HKK26_RS00190': 'Synpcc7942_B2623',\n",
    "            'HKK26_RS00235': 'Synpcc7942_B2655', \n",
    "            'HKK26_RS00240': 'Synpcc7942_B2632',\n",
    "            'HKK26_RS00020': 'Synpcc7942_B2635'}\n",
    "\n",
    "circadian_genes_up = ann[(ann.old_locus_tag.isin(circadian_genes['Gene ID'])) | (ann.index.isin(map_dict.keys()))]\n",
    "print(f\"Number of genes mapped to annotation: {circadian_genes_up.shape[0]}\")\n",
    "circadian_genes_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "1b0ffbcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_lowlight_dawn_0.5h</th>\n",
       "      <th>wt_lowlight_dawn_2h</th>\n",
       "      <th>wt_lowlight_dawn_4h</th>\n",
       "      <th>wt_lowlight_dawn_6h</th>\n",
       "      <th>wt_lowlight_dawn_8h</th>\n",
       "      <th>wt_lowlight_dawn_9h</th>\n",
       "      <th>wt_lowlight_dawn_10h</th>\n",
       "      <th>wt_lowlight_dawn_12h</th>\n",
       "      <th>wt_clearday_dawn_2h</th>\n",
       "      <th>wt_clearday_dawn_4h</th>\n",
       "      <th>...</th>\n",
       "      <th>wt_highlight_pulse_9.25h</th>\n",
       "      <th>wt_highlight_pulse_9.5h</th>\n",
       "      <th>wt_highlight_pulse_10h</th>\n",
       "      <th>wt_shade_pulse_8h</th>\n",
       "      <th>wt_shade_pulse_8.25h</th>\n",
       "      <th>wt_shade_pulse_8.5h</th>\n",
       "      <th>wt_shade_pulse_9h</th>\n",
       "      <th>wt_shade_pulse_9.25h</th>\n",
       "      <th>wt_shade_pulse_9.5h</th>\n",
       "      <th>wt_shade_pulse_10h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6.115376</td>\n",
       "      <td>0.713443</td>\n",
       "      <td>5.495462</td>\n",
       "      <td>5.273049</td>\n",
       "      <td>4.442937</td>\n",
       "      <td>3.272571</td>\n",
       "      <td>2.836849</td>\n",
       "      <td>3.316532</td>\n",
       "      <td>-0.531633</td>\n",
       "      <td>3.782513</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.225281</td>\n",
       "      <td>0.573574</td>\n",
       "      <td>2.066296</td>\n",
       "      <td>2.203895</td>\n",
       "      <td>0.303699</td>\n",
       "      <td>0.656124</td>\n",
       "      <td>1.037911</td>\n",
       "      <td>4.006790</td>\n",
       "      <td>2.951316</td>\n",
       "      <td>1.298457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.248676</td>\n",
       "      <td>-0.937995</td>\n",
       "      <td>-5.065755</td>\n",
       "      <td>-10.140796</td>\n",
       "      <td>-12.870681</td>\n",
       "      <td>-11.926588</td>\n",
       "      <td>-11.279455</td>\n",
       "      <td>-10.108009</td>\n",
       "      <td>0.518376</td>\n",
       "      <td>-2.770383</td>\n",
       "      <td>...</td>\n",
       "      <td>-20.002566</td>\n",
       "      <td>-19.826402</td>\n",
       "      <td>-15.607871</td>\n",
       "      <td>-12.388667</td>\n",
       "      <td>-19.352622</td>\n",
       "      <td>-20.212467</td>\n",
       "      <td>-18.382456</td>\n",
       "      <td>-13.501904</td>\n",
       "      <td>-13.467382</td>\n",
       "      <td>-14.270156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.133630</td>\n",
       "      <td>0.480528</td>\n",
       "      <td>-2.304670</td>\n",
       "      <td>-5.256915</td>\n",
       "      <td>-5.876160</td>\n",
       "      <td>-4.045070</td>\n",
       "      <td>-3.457322</td>\n",
       "      <td>-2.933155</td>\n",
       "      <td>6.389302</td>\n",
       "      <td>3.143030</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.286208</td>\n",
       "      <td>-6.702305</td>\n",
       "      <td>-6.648800</td>\n",
       "      <td>-1.525668</td>\n",
       "      <td>-6.582757</td>\n",
       "      <td>-9.393851</td>\n",
       "      <td>-9.860173</td>\n",
       "      <td>-6.932270</td>\n",
       "      <td>-2.141695</td>\n",
       "      <td>-1.488718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.016743</td>\n",
       "      <td>-0.677561</td>\n",
       "      <td>1.119265</td>\n",
       "      <td>3.072089</td>\n",
       "      <td>4.615046</td>\n",
       "      <td>5.071194</td>\n",
       "      <td>4.112884</td>\n",
       "      <td>4.287350</td>\n",
       "      <td>0.212489</td>\n",
       "      <td>2.811278</td>\n",
       "      <td>...</td>\n",
       "      <td>4.306938</td>\n",
       "      <td>0.986831</td>\n",
       "      <td>4.312314</td>\n",
       "      <td>1.790151</td>\n",
       "      <td>7.804377</td>\n",
       "      <td>5.035021</td>\n",
       "      <td>5.449592</td>\n",
       "      <td>0.279242</td>\n",
       "      <td>1.924907</td>\n",
       "      <td>3.072254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.744702</td>\n",
       "      <td>3.704905</td>\n",
       "      <td>2.669867</td>\n",
       "      <td>2.326103</td>\n",
       "      <td>2.262399</td>\n",
       "      <td>3.160924</td>\n",
       "      <td>1.531659</td>\n",
       "      <td>3.776586</td>\n",
       "      <td>2.761959</td>\n",
       "      <td>3.731732</td>\n",
       "      <td>...</td>\n",
       "      <td>2.703554</td>\n",
       "      <td>2.016426</td>\n",
       "      <td>2.577973</td>\n",
       "      <td>3.315887</td>\n",
       "      <td>2.580680</td>\n",
       "      <td>3.114974</td>\n",
       "      <td>2.111801</td>\n",
       "      <td>0.803972</td>\n",
       "      <td>0.145796</td>\n",
       "      <td>1.672850</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wt_lowlight_dawn_0.5h  wt_lowlight_dawn_2h  wt_lowlight_dawn_4h  \\\n",
       "0               6.115376             0.713443             5.495462   \n",
       "1               0.248676            -0.937995            -5.065755   \n",
       "2               3.133630             0.480528            -2.304670   \n",
       "3               1.016743            -0.677561             1.119265   \n",
       "4               1.744702             3.704905             2.669867   \n",
       "\n",
       "   wt_lowlight_dawn_6h  wt_lowlight_dawn_8h  wt_lowlight_dawn_9h  \\\n",
       "0             5.273049             4.442937             3.272571   \n",
       "1           -10.140796           -12.870681           -11.926588   \n",
       "2            -5.256915            -5.876160            -4.045070   \n",
       "3             3.072089             4.615046             5.071194   \n",
       "4             2.326103             2.262399             3.160924   \n",
       "\n",
       "   wt_lowlight_dawn_10h  wt_lowlight_dawn_12h  wt_clearday_dawn_2h  \\\n",
       "0              2.836849              3.316532            -0.531633   \n",
       "1            -11.279455            -10.108009             0.518376   \n",
       "2             -3.457322             -2.933155             6.389302   \n",
       "3              4.112884              4.287350             0.212489   \n",
       "4              1.531659              3.776586             2.761959   \n",
       "\n",
       "   wt_clearday_dawn_4h  ...  wt_highlight_pulse_9.25h  \\\n",
       "0             3.782513  ...                 -0.225281   \n",
       "1            -2.770383  ...                -20.002566   \n",
       "2             3.143030  ...                 -3.286208   \n",
       "3             2.811278  ...                  4.306938   \n",
       "4             3.731732  ...                  2.703554   \n",
       "\n",
       "   wt_highlight_pulse_9.5h  wt_highlight_pulse_10h  wt_shade_pulse_8h  \\\n",
       "0                 0.573574                2.066296           2.203895   \n",
       "1               -19.826402              -15.607871         -12.388667   \n",
       "2                -6.702305               -6.648800          -1.525668   \n",
       "3                 0.986831                4.312314           1.790151   \n",
       "4                 2.016426                2.577973           3.315887   \n",
       "\n",
       "   wt_shade_pulse_8.25h  wt_shade_pulse_8.5h  wt_shade_pulse_9h  \\\n",
       "0              0.303699             0.656124           1.037911   \n",
       "1            -19.352622           -20.212467         -18.382456   \n",
       "2             -6.582757            -9.393851          -9.860173   \n",
       "3              7.804377             5.035021           5.449592   \n",
       "4              2.580680             3.114974           2.111801   \n",
       "\n",
       "   wt_shade_pulse_9.25h  wt_shade_pulse_9.5h  wt_shade_pulse_10h  \n",
       "0              4.006790             2.951316            1.298457  \n",
       "1            -13.501904           -13.467382          -14.270156  \n",
       "2             -6.932270            -2.141695           -1.488718  \n",
       "3              0.279242             1.924907            3.072254  \n",
       "4              0.803972             0.145796            1.672850  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load difference in gene expression data for TFs (1. Input into ICA)\n",
    "light_pert_samples = light_pert[['SRA', 'Light', 'Hour_Sin', 'Hour_Cos']].set_index('SRA').T\n",
    "light_pert_samples\n",
    "\n",
    "oshea_samples_logTPM_norm_TFs = logTPM_norm.loc[pred_TFs.TFs, light_pert_samples.columns]\n",
    "oshea_samples_logTPM_norm_TFs.index = oshea_samples_logTPM_norm_TFs.index.map('TF_{}'.format)\n",
    "oshea_samples_logTPM_norm_TFs\n",
    "\n",
    "light_pert_conditions = light_pert[['Condition_pi', 'Light', 'Hour_Sin', 'Hour_Cos']\n",
    "                                  ].drop_duplicates().set_index('Condition_pi').T.drop(columns='wt_clearday_dawn_0.5h')\n",
    "\n",
    "#Load log2FC data (2. Output from DESeq2)\n",
    "oshea_samples_log2FC_TFs = log2FC.loc[pred_TFs.TFs, light_pert_conditions.columns]\n",
    "oshea_samples_log2FC_TFs.index = oshea_samples_log2FC_TFs.index.map('TF_{}'.format)\n",
    "\n",
    "# Modulon response by sample\n",
    "oshea_IC_sample = A.loc[:, light_pert_samples.columns]\n",
    "\n",
    "## Average the modulon resposne across conditions \n",
    "## wt_clearday_dawn_0.5h removed as it was used as the base condition for all differential expression comparisons\n",
    "\n",
    "avg_A = []\n",
    "for sample, group in light_pert.groupby('Condition_pi'):\n",
    "    avg = oshea_IC_sample[group.SRA].mean(axis=1)\n",
    "    avg_A.append(avg)\n",
    "    \n",
    "avg_A_df = pd.concat(avg_A, axis=1)\n",
    "avg_A_df.columns = light_pert.groupby('Condition_pi').groups.keys()\n",
    "avg_A_df = avg_A_df.drop('wt_clearday_dawn_0.5h', axis=1)\n",
    "avg_A_df = avg_A_df.reindex(light_pert_conditions.columns.values, axis=1)\n",
    "avg_A_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3e73f75",
   "metadata": {},
   "source": [
    "# Module model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5d6bd91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Known Regulators identified in circadian phase: ['RpoD6' 'RpoD5' 'SigF2' 'CcmR' 'kaiC' 'pex']\n",
      "Inclusion of known regulators: ['RpaA' 'RpaB' 'RpoD6' 'RpoD5' 'SigF2' 'NtcA' 'CcmR' 'HrcA' 'kaiC' 'rre1'\n",
      " 'pex']\n"
     ]
    }
   ],
   "source": [
    "circadian_tfs = set(circadian_genes_up.index).intersection(set(pred_TFs.TFs))\n",
    "add_regs = ['RpaA', 'RpaB', 'rre1', 'HrcA', 'NtcA']\n",
    "add_regs_locus_tags = regs[regs.Regulator_Name.isin(add_regs)].regulatoryGene.unique()\n",
    "\n",
    "\n",
    "print(f'Known Regulators identified in circadian phase: {regs[regs.regulatoryGene.isin(circadian_tfs)].Regulator_Name.unique()}')\n",
    "print(f'Inclusion of known regulators: {regs[regs.regulatoryGene.isin(circadian_tfs.union(set(add_regs_locus_tags)))].Regulator_Name.unique()}')\n",
    "\n",
    "circ_log2FC_TFs = oshea_samples_log2FC_TFs.loc[list([\"TF_\" + TF for TF in circadian_tfs.union(set(add_regs_locus_tags))]),:]\n",
    "circ_log2FC_TFs = circ_log2FC_TFs.reindex(tf_order[0].values)\n",
    "construct_cellbox_input(light_pert_conditions, circ_log2FC_TFs, avg_A_df, '../Execution_files/Cellbox_Module')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10487b21",
   "metadata": {},
   "source": [
    "# Gene model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9994be01",
   "metadata": {},
   "outputs": [],
   "source": [
    "circ_log2FC_genes = log2FC[log2FC.index.isin(circadian_genes_up.index)]\n",
    "\n",
    "construct_cellbox_input(light_pert_conditions, circ_log2FC_TFs, circ_log2FC_genes, \n",
    "                        \"../Execution_files/Cellbox_Gene\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pimlomics_conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
